latex_directive_completions.py

```diff
@@ -0,0 +1,222 @@
+    if not view.file_name():
+        return []
+
+    directory, base_name = os.path.split(view.file_name())
+    exts = get_setting("tex_file_exts", view=view)
+
+    def is_tex_file(file_name):
+        return any(file_name.endswith(e) for e in exts)
+
+    def list_tex_files(dir_path):
+        return (f for f in os.listdir(dir_path) if is_tex_file(f))
+
+    tex_files = [
+        "./" + f for f in list_tex_files(directory) if not base_name == f
+    ]
+
+    # search up to 3 empty folders for tex files
+    search_threshold = 3
+    search_miss = 0
+
+    path = "../"
+
+    while search_miss < search_threshold:
+        parent = os.path.abspath(os.path.join(directory, path))
+
+        parent_tex_files = list(path + s for s in list_tex_files(parent))
+        tex_files.extend(parent_tex_files)
+
+        if not parent_tex_files:
+            search_miss += 1
+        else:
+            search_miss = 0
+
+        path += "../"
+        # ensure we are not going into an infinite loop if someone
+        # is working on the root directory
+        if parent == os.path.abspath(os.path.join(directory, path)):
+            break
+
+    len_prefix = len([v for v in value if v in [".", "/"]])
+    tex_files = [
+        s for s in tex_files
+        if s.startswith(value) or
+        (not len_prefix and s.startswith("./" + value))
+    ]
+    if ac:
+        comp = [(s + "\ttex-file", s[len_prefix:]) for s in tex_files]
+    else:
+        comp = [
+            [s, os.path.abspath(os.path.join(directory, s))] for s in tex_files
+        ], tex_files
+
+    return comp
+
+
+def _directive_spellcheck_completions(view, value, ac=True):
+    user_sc = get_setting("tex_spellcheck_paths", view=view, default={})
+    locales = sorted(user_sc.keys())
+
+    locales.extend(installed_locales)
+
+    def get_locale(loc):
+        try:
+            loc = detect_spellcheck.normalize_locale(loc)
+            dic = user_sc.get(loc) or detect_spellcheck.get_dict_path(loc)
+            if ac:
+                _, dic = os.path.split(dic)
+            elif dic.startswith("Packages/"):
+                dic = dic[len("Packages/"):]
+        except:
+            dic = "locale"
+        return dic
+    locales = [
+        loc
+        for loc in map(_prettify_locale, locales)
+        if loc.startswith(value)
+    ]
+
+    if ac:
+        comp = [
+            ("{0}\t{1}".format(loc, get_locale(loc)), loc)
+            for loc in locales
+        ]
+    else:
+        comp = [[loc, get_locale(loc)] for loc in locales], locales
+    return comp
+
+
+def _directive_program_completions(view, value, ac=True):
+    engines = [
+        "pdflatex", "xelatex", "lualatex", "pdftex", "xetex", "luatex"
+    ]
+    engines = [e for e in engines if e.startswith(value)]
+    if ac:
+        comp = [(e + "\ttex-program", e) for e in engines]
+    else:
+        comp = engines, engines
+    return comp
+
+
+_LINE_RE = re.compile(
```

Comment:
This definition doesn't quite match the definition it `get_tex_root`. In particular, `get_tex_root` supports multiple percent signs (I've come across documents with `%%% !TEX root = ...`, etc.) and it doesn't support space before the percent sign. I'd actually be happy for us to add support for spaces before the `%`; it just should be handled consistently.